{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "16a4894c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/mnt/nzh/nzhanglab/project/shuang/scATAC/comparison_methods/scglue',\n",
       " '/usr/local/python-3.8.5/lib/python38.zip',\n",
       " '/usr/local/python-3.8.5/lib/python3.8',\n",
       " '/usr/local/python-3.8.5/lib/python3.8/lib-dynload',\n",
       " '',\n",
       " '/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages',\n",
       " '/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/stereoscope-0.3.1-py3.8.egg']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.path\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bb076ecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('/home/mnt/nzh/nzhanglab/project/shuang/miniconda3/envs/scglue2/lib/python3.8/site-packages')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d09248fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import networkx as nx\n",
    "import scanpy as sc\n",
    "import scglue\n",
    "from matplotlib import rcParams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d3d58844",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The current date and time is 18:29:27\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "time_now = datetime.now()\n",
    "current_time = time_now.strftime(\"%H:%M:%S\")\n",
    "\n",
    "print(\"The current date and time is\", current_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "66feeae7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2500 × 36601\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'celltype', 'domain'\n",
       "    var: 'features'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna = anndata.read_h5ad(\"retina/retina_RNA.2000.h5ad\")\n",
    "rna\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3658fc27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2500 × 325401\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'celltype', 'domain'\n",
       "    var: 'features'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac = anndata.read_h5ad(\"retina/retina_ATAC.2000.h5ad\")\n",
    "atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4bb811c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<2500x36601 sparse matrix of type '<class 'numpy.float64'>'\n",
       " \twith 4370242 stored elements in Compressed Sparse Row format>,\n",
       " array([1., 1., 2., ..., 1., 1., 1.]))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.X, rna.X.data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab8776a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.layers[\"counts\"] = rna.X.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "340f5ef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(rna, n_top_genes=2000, flavor=\"seurat_v3\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c899b005",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(rna)\n",
    "sc.pp.log1p(rna)\n",
    "sc.pp.scale(rna)\n",
    "sc.tl.pca(rna, n_comps=100, svd_solver=\"auto\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "198e242b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<2500x325401 sparse matrix of type '<class 'numpy.float64'>'\n",
       " \twith 22494323 stored elements in Compressed Sparse Row format>,\n",
       " array([2., 2., 1., ..., 1., 2., 2.]))"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.X, atac.X.data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "62f107df",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mnt/nzh/nzhanglab/project/shuang/miniconda3/envs/scglue2/lib/python3.8/site-packages/scglue/num.py:289: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  idf = X.shape[0] / X.sum(axis=0)\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/sklearn/utils/extmath.py:368: FutureWarning: If 'random_state' is not supplied, the current default is to use 0 as a fixed seed. This will change to  None in version 1.2 leading to non-deterministic results that better reflect nature of the randomized_svd solver. If you want to silence this warning, set 'random_state' to an integer seed or to None explicitly depending if you want your code to be deterministic or not.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "scglue.data.lsi(atac, n_components=100, n_iter=15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c3924ab3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(atac, use_rep=\"X_lsi\", metric=\"cosine\")\n",
    "sc.tl.umap(atac)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4cb477e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>features</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>highly_variable_rank</th>\n",
       "      <th>means</th>\n",
       "      <th>variances</th>\n",
       "      <th>variances_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024</td>\n",
       "      <td>0.002395</td>\n",
       "      <td>0.876998</td>\n",
       "      <td>0.001088</td>\n",
       "      <td>0.024788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0056</td>\n",
       "      <td>0.005571</td>\n",
       "      <td>0.825701</td>\n",
       "      <td>0.003545</td>\n",
       "      <td>0.054385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                features  highly_variable  highly_variable_rank   means  \\\n",
       "MIR1302-2HG  MIR1302-2HG            False                   NaN  0.0024   \n",
       "FAM138A          FAM138A            False                   NaN  0.0000   \n",
       "OR4F5              OR4F5            False                   NaN  0.0000   \n",
       "AL627309.1    AL627309.1            False                   NaN  0.0056   \n",
       "AL627309.3    AL627309.3            False                   NaN  0.0000   \n",
       "\n",
       "             variances  variances_norm      mean       std  \n",
       "MIR1302-2HG   0.002395        0.876998  0.001088  0.024788  \n",
       "FAM138A       0.000000        0.000000  0.000000  1.000000  \n",
       "OR4F5         0.000000        0.000000  0.000000  1.000000  \n",
       "AL627309.1    0.005571        0.825701  0.003545  0.054385  \n",
       "AL627309.3    0.000000        0.000000  0.000000  1.000000  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.var.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "96bcd7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.var_names.name = 'genes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "522af65c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['MIR1302-2HG', 'FAM138A', 'OR4F5', 'AL627309.1', 'AL627309.3',\n",
       "       'AL627309.2', 'AL627309.5', 'AL627309.4', 'AP006222.2', 'AL732372.1',\n",
       "       ...\n",
       "       'AC133551.1', 'AC136612.1', 'AC136616.1', 'AC136616.3', 'AC136616.2',\n",
       "       'AC141272.1', 'AC023491.2', 'AC007325.1', 'AC007325.4', 'AC007325.2'],\n",
       "      dtype='object', name='genes', length=36601)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e714497d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29553.0</td>\n",
       "      <td>31109.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>chr1</td>\n",
       "      <td>34553.0</td>\n",
       "      <td>36081.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418.0</td>\n",
       "      <td>71585.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chrom  chromStart  chromEnd\n",
       "genes                                  \n",
       "MIR1302-2HG  chr1     29553.0   31109.0\n",
       "FAM138A      chr1     34553.0   36081.0\n",
       "OR4F5        chr1     65418.0   71585.0\n",
       "AL627309.1    NaN         NaN       NaN\n",
       "AL627309.3    NaN         NaN       NaN"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scglue.data.get_gene_annotation(\n",
    "    rna, gtf=\"gencode.v40.chr_patch_hapl_scaff.annotation.gtf.gz\",\n",
    "    gtf_by=\"gene_name\"\n",
    ")\n",
    "rna.var.loc[:, [\"chrom\", \"chromStart\", \"chromEnd\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5921402c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>features</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>highly_variable_rank</th>\n",
       "      <th>means</th>\n",
       "      <th>variances</th>\n",
       "      <th>variances_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>...</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>itemRgb</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>blockStarts</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024</td>\n",
       "      <td>0.002395</td>\n",
       "      <td>0.876998</td>\n",
       "      <td>0.001088</td>\n",
       "      <td>0.024788</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29553.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ncRNA_host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>34553.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>HGNC:32334</td>\n",
       "      <td>OTTHUMG00000000960.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65418.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>OTTHUMG00000001094.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0056</td>\n",
       "      <td>0.005571</td>\n",
       "      <td>0.825701</td>\n",
       "      <td>0.003545</td>\n",
       "      <td>0.054385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC141272.1</th>\n",
       "      <td>AC141272.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "      <td>AC023491.2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "      <td>AC007325.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>0.999718</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>0.004923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0228</td>\n",
       "      <td>0.027091</td>\n",
       "      <td>0.906645</td>\n",
       "      <td>0.009099</td>\n",
       "      <td>0.076471</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0.001598</td>\n",
       "      <td>0.910282</td>\n",
       "      <td>0.000581</td>\n",
       "      <td>0.016496</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                features  highly_variable  highly_variable_rank   means  \\\n",
       "genes                                                                     \n",
       "MIR1302-2HG  MIR1302-2HG            False                   NaN  0.0024   \n",
       "FAM138A          FAM138A            False                   NaN  0.0000   \n",
       "OR4F5              OR4F5            False                   NaN  0.0000   \n",
       "AL627309.1    AL627309.1            False                   NaN  0.0056   \n",
       "AL627309.3    AL627309.3            False                   NaN  0.0000   \n",
       "...                  ...              ...                   ...     ...   \n",
       "AC141272.1    AC141272.1            False                   NaN  0.0000   \n",
       "AC023491.2    AC023491.2            False                   NaN  0.0000   \n",
       "AC007325.1    AC007325.1            False                   NaN  0.0004   \n",
       "AC007325.4    AC007325.4            False                   NaN  0.0228   \n",
       "AC007325.2    AC007325.2            False                   NaN  0.0016   \n",
       "\n",
       "             variances  variances_norm      mean       std chrom  chromStart  \\\n",
       "genes                                                                          \n",
       "MIR1302-2HG   0.002395        0.876998  0.001088  0.024788  chr1     29553.0   \n",
       "FAM138A       0.000000        0.000000  0.000000  1.000000  chr1     34553.0   \n",
       "OR4F5         0.000000        0.000000  0.000000  1.000000  chr1     65418.0   \n",
       "AL627309.1    0.005571        0.825701  0.003545  0.054385   NaN         NaN   \n",
       "AL627309.3    0.000000        0.000000  0.000000  1.000000   NaN         NaN   \n",
       "...                ...             ...       ...       ...   ...         ...   \n",
       "AC141272.1    0.000000        0.000000  0.000000  1.000000   NaN         NaN   \n",
       "AC023491.2    0.000000        0.000000  0.000000  1.000000   NaN         NaN   \n",
       "AC007325.1    0.000400        0.999718  0.000098  0.004923   NaN         NaN   \n",
       "AC007325.4    0.027091        0.906645  0.009099  0.076471   NaN         NaN   \n",
       "AC007325.2    0.001598        0.910282  0.000581  0.016496   NaN         NaN   \n",
       "\n",
       "             ...  thickEnd itemRgb blockCount blockSizes blockStarts  \\\n",
       "genes        ...                                                       \n",
       "MIR1302-2HG  ...         .       .          .          .           .   \n",
       "FAM138A      ...         .       .          .          .           .   \n",
       "OR4F5        ...         .       .          .          .           .   \n",
       "AL627309.1   ...       NaN     NaN        NaN        NaN         NaN   \n",
       "AL627309.3   ...       NaN     NaN        NaN        NaN         NaN   \n",
       "...          ...       ...     ...        ...        ...         ...   \n",
       "AC141272.1   ...       NaN     NaN        NaN        NaN         NaN   \n",
       "AC023491.2   ...       NaN     NaN        NaN        NaN         NaN   \n",
       "AC007325.1   ...       NaN     NaN        NaN        NaN         NaN   \n",
       "AC007325.4   ...       NaN     NaN        NaN        NaN         NaN   \n",
       "AC007325.2   ...       NaN     NaN        NaN        NaN         NaN   \n",
       "\n",
       "                       gene_id       gene_type     hgnc_id  \\\n",
       "genes                                                        \n",
       "MIR1302-2HG  ENSG00000243485.5          lncRNA  HGNC:52482   \n",
       "FAM138A      ENSG00000237613.2          lncRNA  HGNC:32334   \n",
       "OR4F5        ENSG00000186092.7  protein_coding  HGNC:14825   \n",
       "AL627309.1                 NaN             NaN         NaN   \n",
       "AL627309.3                 NaN             NaN         NaN   \n",
       "...                        ...             ...         ...   \n",
       "AC141272.1                 NaN             NaN         NaN   \n",
       "AC023491.2                 NaN             NaN         NaN   \n",
       "AC007325.1                 NaN             NaN         NaN   \n",
       "AC007325.4                 NaN             NaN         NaN   \n",
       "AC007325.2                 NaN             NaN         NaN   \n",
       "\n",
       "                      havana_gene         tag  \n",
       "genes                                          \n",
       "MIR1302-2HG  OTTHUMG00000000959.2  ncRNA_host  \n",
       "FAM138A      OTTHUMG00000000960.1         NaN  \n",
       "OR4F5        OTTHUMG00000001094.4         NaN  \n",
       "AL627309.1                    NaN         NaN  \n",
       "AL627309.3                    NaN         NaN  \n",
       "...                           ...         ...  \n",
       "AC141272.1                    NaN         NaN  \n",
       "AC023491.2                    NaN         NaN  \n",
       "AC007325.1                    NaN         NaN  \n",
       "AC007325.4                    NaN         NaN  \n",
       "AC007325.2                    NaN         NaN  \n",
       "\n",
       "[36601 rows x 25 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "17093447",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = rna.var\n",
    "#print(df.dropna(subset=['chrom'])[\"features\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e6bcd258",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna2 = rna\n",
    "rna2 = rna2[:,df.dropna(subset=['chrom'])[\"features\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1682057c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29553.0</td>\n",
       "      <td>31109.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>chr1</td>\n",
       "      <td>34553.0</td>\n",
       "      <td>36081.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418.0</td>\n",
       "      <td>71585.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450739.0</td>\n",
       "      <td>451678.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685715.0</td>\n",
       "      <td>686654.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>10759.0</td>\n",
       "      <td>12137.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND5</th>\n",
       "      <td>chrM</td>\n",
       "      <td>12336.0</td>\n",
       "      <td>14148.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14148.0</td>\n",
       "      <td>14673.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14746.0</td>\n",
       "      <td>15887.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAFIP</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>53593.0</td>\n",
       "      <td>115055.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23582 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  chrom  chromStart  chromEnd\n",
       "genes                                        \n",
       "MIR1302-2HG        chr1     29553.0   31109.0\n",
       "FAM138A            chr1     34553.0   36081.0\n",
       "OR4F5              chr1     65418.0   71585.0\n",
       "OR4F29             chr1    450739.0  451678.0\n",
       "OR4F16             chr1    685715.0  686654.0\n",
       "...                 ...         ...       ...\n",
       "MT-ND4             chrM     10759.0   12137.0\n",
       "MT-ND5             chrM     12336.0   14148.0\n",
       "MT-ND6             chrM     14148.0   14673.0\n",
       "MT-CYB             chrM     14746.0   15887.0\n",
       "MAFIP        GL000194.1     53593.0  115055.0\n",
       "\n",
       "[23582 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna2.var.loc[:, [\"chrom\", \"chromStart\", \"chromEnd\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7aaf5389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['MIR1302-2HG', 'FAM138A', 'OR4F5', 'OR4F29', 'OR4F16', 'LINC01409',\n",
       "       'FAM87B', 'LINC01128', 'LINC00115', 'FAM41C',\n",
       "       ...\n",
       "       'MT-ATP8', 'MT-ATP6', 'MT-CO3', 'MT-ND3', 'MT-ND4L', 'MT-ND4', 'MT-ND5',\n",
       "       'MT-ND6', 'MT-CYB', 'MAFIP'],\n",
       "      dtype='object', name='genes', length=23582)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna2.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e796197a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rna2.var.loc[:, [\"chrom\", \"chromStart\", \"chromEnd\"]].to_csv('e18mouse/rnavar2.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "277b4fd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr1-794327-794827', 'chr1-794856-795356', 'chr1-804383-804883',\n",
       "       'chr1-817118-817618', 'chr1-819862-820362'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.var_names[:5]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "699d2344",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>features</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1-794327-794827</th>\n",
       "      <td>chr1-794327-794827</td>\n",
       "      <td>chr1</td>\n",
       "      <td>794327</td>\n",
       "      <td>794827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-794856-795356</th>\n",
       "      <td>chr1-794856-795356</td>\n",
       "      <td>chr1</td>\n",
       "      <td>794856</td>\n",
       "      <td>795356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-804383-804883</th>\n",
       "      <td>chr1-804383-804883</td>\n",
       "      <td>chr1</td>\n",
       "      <td>804383</td>\n",
       "      <td>804883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-817118-817618</th>\n",
       "      <td>chr1-817118-817618</td>\n",
       "      <td>chr1</td>\n",
       "      <td>817118</td>\n",
       "      <td>817618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-819862-820362</th>\n",
       "      <td>chr1-819862-820362</td>\n",
       "      <td>chr1</td>\n",
       "      <td>819862</td>\n",
       "      <td>820362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              features chrom chromStart chromEnd\n",
       "chr1-794327-794827  chr1-794327-794827  chr1     794327   794827\n",
       "chr1-794856-795356  chr1-794856-795356  chr1     794856   795356\n",
       "chr1-804383-804883  chr1-804383-804883  chr1     804383   804883\n",
       "chr1-817118-817618  chr1-817118-817618  chr1     817118   817618\n",
       "chr1-819862-820362  chr1-819862-820362  chr1     819862   820362"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "split = atac.var_names.str.split(r\"[:-]\")\n",
    "atac.var[\"chrom\"] = split.map(lambda x: x[0])\n",
    "atac.var[\"chromStart\"] = split.map(lambda x: x[1])\n",
    "atac.var[\"chromEnd\"] = split.map(lambda x: x[2])\n",
    "atac.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2a8ba71d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/python-3.8.5/lib/python3.8/subprocess.py:838: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stdin = io.open(p2cwrite, 'wb', bufsize)\n",
      "/usr/local/python-3.8.5/lib/python3.8/subprocess.py:844: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stdout = io.open(c2pread, 'rb', bufsize)\n",
      "/usr/local/python-3.8.5/lib/python3.8/subprocess.py:849: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stderr = io.open(errread, 'rb', bufsize)\n",
      "/usr/local/python-3.8.5/lib/python3.8/subprocess.py:838: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stdin = io.open(p2cwrite, 'wb', bufsize)\n",
      "/usr/local/python-3.8.5/lib/python3.8/subprocess.py:844: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stdout = io.open(c2pread, 'rb', bufsize)\n",
      "/usr/local/python-3.8.5/lib/python3.8/subprocess.py:849: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stderr = io.open(errread, 'rb', bufsize)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ea93ef4841f04c559741cddbb72405bd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/23582 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<networkx.classes.multidigraph.MultiDiGraph at 0x2b7bdbcfd310>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph = scglue.genomics.rna_anchored_prior_graph(rna2, atac)\n",
    "graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c6efc816",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(348983, 874335)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph.number_of_nodes(), graph.number_of_edges()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d0af4b8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, True)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Graph node covers all omic features\n",
    "all(graph.has_node(gene) for gene in rna2.var_names), \\\n",
    "all(graph.has_node(peak) for peak in atac.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "430475d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('LINC01409', 'chr1-794327-794827', 0): {'dist': 0, 'weight': 1.0, 'sign': 1, 'type': 'fwd'}\n",
      "('LINC01409', 'chr1-794856-795356', 0): {'dist': 0, 'weight': 1.0, 'sign': 1, 'type': 'fwd'}\n",
      "('LINC01409', 'chr1-804383-804883', 0): {'dist': 0, 'weight': 1.0, 'sign': 1, 'type': 'fwd'}\n",
      "('LINC01409', 'LINC01409', 0): {'weight': 1.0, 'sign': 1, 'type': 'loop'}\n",
      "('chr1-794327-794827', 'LINC01409', 0): {'dist': 0, 'weight': 1.0, 'sign': 1, 'type': 'rev'}\n"
     ]
    }
   ],
   "source": [
    "# Edge attributes contain weights and signs\n",
    "for _, e in zip(range(5), graph.edges):\n",
    "    print(f\"{e}: {graph.edges[e]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0cea7c74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, True)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Each node has a self-loop\n",
    "all(graph.has_edge(gene, gene) for gene in rna2.var_names), \\\n",
    "all(graph.has_edge(peak, peak) for peak in atac.var_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3634ae80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Graph is symmetric\n",
    "all(graph.has_edge(j, i) for i, j, _ in graph.edges)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1ec361a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>features</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>highly_variable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1-794327-794827</th>\n",
       "      <td>chr1-794327-794827</td>\n",
       "      <td>chr1</td>\n",
       "      <td>794327</td>\n",
       "      <td>794827</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-794856-795356</th>\n",
       "      <td>chr1-794856-795356</td>\n",
       "      <td>chr1</td>\n",
       "      <td>794856</td>\n",
       "      <td>795356</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-804383-804883</th>\n",
       "      <td>chr1-804383-804883</td>\n",
       "      <td>chr1</td>\n",
       "      <td>804383</td>\n",
       "      <td>804883</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-817118-817618</th>\n",
       "      <td>chr1-817118-817618</td>\n",
       "      <td>chr1</td>\n",
       "      <td>817118</td>\n",
       "      <td>817618</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-819862-820362</th>\n",
       "      <td>chr1-819862-820362</td>\n",
       "      <td>chr1</td>\n",
       "      <td>819862</td>\n",
       "      <td>820362</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              features chrom chromStart chromEnd  \\\n",
       "chr1-794327-794827  chr1-794327-794827  chr1     794327   794827   \n",
       "chr1-794856-795356  chr1-794856-795356  chr1     794856   795356   \n",
       "chr1-804383-804883  chr1-804383-804883  chr1     804383   804883   \n",
       "chr1-817118-817618  chr1-817118-817618  chr1     817118   817618   \n",
       "chr1-819862-820362  chr1-819862-820362  chr1     819862   820362   \n",
       "\n",
       "                    highly_variable  \n",
       "chr1-794327-794827            False  \n",
       "chr1-794856-795356            False  \n",
       "chr1-804383-804883            False  \n",
       "chr1-817118-817618            False  \n",
       "chr1-819862-820362            False  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.var.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5d28c88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = rna2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "521ab1dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>features</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>highly_variable_rank</th>\n",
       "      <th>means</th>\n",
       "      <th>variances</th>\n",
       "      <th>variances_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>...</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>itemRgb</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>blockStarts</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024</td>\n",
       "      <td>0.002395</td>\n",
       "      <td>0.876998</td>\n",
       "      <td>0.001088</td>\n",
       "      <td>0.024788</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29553.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ncRNA_host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>34553.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>HGNC:32334</td>\n",
       "      <td>OTTHUMG00000000960.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65418.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>OTTHUMG00000001094.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F29</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>450739.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:31275</td>\n",
       "      <td>OTTHUMG00000002860.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F16</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>685715.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:15079</td>\n",
       "      <td>OTTHUMG00000002581.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>0.150184</td>\n",
       "      <td>0.746812</td>\n",
       "      <td>0.065351</td>\n",
       "      <td>0.212767</td>\n",
       "      <td>chrM</td>\n",
       "      <td>10759.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:7459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND5</th>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0240</td>\n",
       "      <td>0.026635</td>\n",
       "      <td>0.843449</td>\n",
       "      <td>0.011898</td>\n",
       "      <td>0.094013</td>\n",
       "      <td>chrM</td>\n",
       "      <td>12336.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:7461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0032</td>\n",
       "      <td>0.003191</td>\n",
       "      <td>0.855367</td>\n",
       "      <td>0.001764</td>\n",
       "      <td>0.036288</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14148.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:7462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0984</td>\n",
       "      <td>0.124767</td>\n",
       "      <td>0.830714</td>\n",
       "      <td>0.046251</td>\n",
       "      <td>0.177238</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14746.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:7427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAFIP</th>\n",
       "      <td>MAFIP</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>0.012661</td>\n",
       "      <td>0.833345</td>\n",
       "      <td>0.005808</td>\n",
       "      <td>0.060494</td>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>53593.0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000274847.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HGNC:31102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23582 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                features  highly_variable  highly_variable_rank   means  \\\n",
       "genes                                                                     \n",
       "MIR1302-2HG  MIR1302-2HG            False                   NaN  0.0024   \n",
       "FAM138A          FAM138A            False                   NaN  0.0000   \n",
       "OR4F5              OR4F5            False                   NaN  0.0000   \n",
       "OR4F29            OR4F29            False                   NaN  0.0000   \n",
       "OR4F16            OR4F16            False                   NaN  0.0000   \n",
       "...                  ...              ...                   ...     ...   \n",
       "MT-ND4            MT-ND4            False                   NaN  0.1260   \n",
       "MT-ND5            MT-ND5            False                   NaN  0.0240   \n",
       "MT-ND6            MT-ND6            False                   NaN  0.0032   \n",
       "MT-CYB            MT-CYB            False                   NaN  0.0984   \n",
       "MAFIP              MAFIP            False                   NaN  0.0120   \n",
       "\n",
       "             variances  variances_norm      mean       std       chrom  \\\n",
       "genes                                                                    \n",
       "MIR1302-2HG   0.002395        0.876998  0.001088  0.024788        chr1   \n",
       "FAM138A       0.000000        0.000000  0.000000  1.000000        chr1   \n",
       "OR4F5         0.000000        0.000000  0.000000  1.000000        chr1   \n",
       "OR4F29        0.000000        0.000000  0.000000  1.000000        chr1   \n",
       "OR4F16        0.000000        0.000000  0.000000  1.000000        chr1   \n",
       "...                ...             ...       ...       ...         ...   \n",
       "MT-ND4        0.150184        0.746812  0.065351  0.212767        chrM   \n",
       "MT-ND5        0.026635        0.843449  0.011898  0.094013        chrM   \n",
       "MT-ND6        0.003191        0.855367  0.001764  0.036288        chrM   \n",
       "MT-CYB        0.124767        0.830714  0.046251  0.177238        chrM   \n",
       "MAFIP         0.012661        0.833345  0.005808  0.060494  GL000194.1   \n",
       "\n",
       "             chromStart  ...  thickEnd itemRgb blockCount blockSizes  \\\n",
       "genes                    ...                                           \n",
       "MIR1302-2HG     29553.0  ...         .       .          .          .   \n",
       "FAM138A         34553.0  ...         .       .          .          .   \n",
       "OR4F5           65418.0  ...         .       .          .          .   \n",
       "OR4F29         450739.0  ...         .       .          .          .   \n",
       "OR4F16         685715.0  ...         .       .          .          .   \n",
       "...                 ...  ...       ...     ...        ...        ...   \n",
       "MT-ND4          10759.0  ...         .       .          .          .   \n",
       "MT-ND5          12336.0  ...         .       .          .          .   \n",
       "MT-ND6          14148.0  ...         .       .          .          .   \n",
       "MT-CYB          14746.0  ...         .       .          .          .   \n",
       "MAFIP           53593.0  ...         .       .          .          .   \n",
       "\n",
       "            blockStarts            gene_id       gene_type     hgnc_id  \\\n",
       "genes                                                                    \n",
       "MIR1302-2HG           .  ENSG00000243485.5          lncRNA  HGNC:52482   \n",
       "FAM138A               .  ENSG00000237613.2          lncRNA  HGNC:32334   \n",
       "OR4F5                 .  ENSG00000186092.7  protein_coding  HGNC:14825   \n",
       "OR4F29                .  ENSG00000284733.2  protein_coding  HGNC:31275   \n",
       "OR4F16                .  ENSG00000284662.2  protein_coding  HGNC:15079   \n",
       "...                 ...                ...             ...         ...   \n",
       "MT-ND4                .  ENSG00000198886.2  protein_coding   HGNC:7459   \n",
       "MT-ND5                .  ENSG00000198786.2  protein_coding   HGNC:7461   \n",
       "MT-ND6                .  ENSG00000198695.2  protein_coding   HGNC:7462   \n",
       "MT-CYB                .  ENSG00000198727.2  protein_coding   HGNC:7427   \n",
       "MAFIP                 .  ENSG00000274847.1  protein_coding  HGNC:31102   \n",
       "\n",
       "                      havana_gene         tag  \n",
       "genes                                          \n",
       "MIR1302-2HG  OTTHUMG00000000959.2  ncRNA_host  \n",
       "FAM138A      OTTHUMG00000000960.1         NaN  \n",
       "OR4F5        OTTHUMG00000001094.4         NaN  \n",
       "OR4F29       OTTHUMG00000002860.3         NaN  \n",
       "OR4F16       OTTHUMG00000002581.3         NaN  \n",
       "...                           ...         ...  \n",
       "MT-ND4                        NaN         NaN  \n",
       "MT-ND5                        NaN         NaN  \n",
       "MT-ND6                        NaN         NaN  \n",
       "MT-CYB                        NaN         NaN  \n",
       "MAFIP                         NaN         NaN  \n",
       "\n",
       "[23582 rows x 25 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "445827af",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna2.raw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c7817f25",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna2.raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5259217",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a1298317",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'chrom' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'score' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'strand' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'thickStart' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'thickEnd' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'itemRgb' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'blockCount' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'blockSizes' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'blockStarts' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'gene_type' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'hgnc_id' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'havana_gene' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'tag' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'chrom' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'chromStart' as categorical\n",
      "/home/stat/shuang91/gpuenv385/lib/python3.8/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'chromEnd' as categorical\n"
     ]
    }
   ],
   "source": [
    "#rna.write(\"e18mouse/rna_preprocessed.h5ad\", compression=\"gzip\")\n",
    "rna2.write(\"retina/rna_preprocessed_2000.h5ad\", compression=\"gzip\")\n",
    "\n",
    "atac.raw = None\n",
    "atac.write(\"retina/atac_preprocessed_2000.h5ad\", compression=\"gzip\")\n",
    "nx.write_graphml(graph, \"retina/prior_2000.graphml.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6f8852ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The current date and time is 18:34:17\n"
     ]
    }
   ],
   "source": [
    "time_now = datetime.now()\n",
    "current_time = time_now.strftime(\"%H:%M:%S\")\n",
    "\n",
    "print(\"The current date and time is\", current_time)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "428a64ae",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "131b471a",
   "metadata": {},
   "source": [
    "#rna.write(\"e18mouse/rna_preprocessed.h5ad\", compression=\"gzip\")\n",
    "rna2.write(\"retina/rna_preprocessed.h5ad\", compression=\"gzip\")\n",
    "\n",
    "atac.raw = None\n",
    "atac.write(\"retina/atac_preprocessed.h5ad\", compression=\"gzip\")\n",
    "nx.write_graphml(graph, \"retina/prior.graphml.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "239fbc93",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
